crates/net/eth-wire/src/types/message.rs

```diff
@@ -22,10 +22,9 @@ pub struct ProtocolMessage {
 
 impl ProtocolMessage {
     /// Create a new ProtocolMessage from a message type and message rlp bytes.
-    pub fn decode_message(
-        message_type: EthMessageID,
-        buf: &mut &[u8],
-    ) -> Result<Self, reth_rlp::DecodeError> {
+    pub fn decode_message(version: u8, buf: &mut &[u8]) -> Result<Self, reth_rlp::DecodeError> {
```

Comment:
so now decode_message depends on the version? because `NewPooledTransactionHashes` format differs?

...
I feel like mixing this together with decode is a bit weird. for example decoding GetNodeData now fails with an rlp error even though it can still be valid rlp.

I really would like to do this in two steps:
decode then validate against version.
wdyt @Rjected

Reply:
yes, decoding should not worry about version details, I think this is another consequence of not supporting multiple capabilities. We don't know what to decode into if we don't know the eth wire protocol version being used. It's ambiguous whether or not we should attempt to decode a `eth/68` pooled hashes message, or a `eth/67` pooled hashes message when we receive a `NewPooledTransactionHashes` message ID.

Reply:
For support multiple capabilities, we must have a branch at some level. It could be just `if`, `match`, `traits`, `module` or whatever. It's needed but I agree that it's not the best way.